typedData/typedData.go

```diff
@@ -0,0 +1,905 @@
+package typedData
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"math/big"
+	"regexp"
+	"slices"
+	"strconv"
+	"strings"
+
+	"github.com/NethermindEth/juno/core/felt"
+	"github.com/NethermindEth/starknet.go/utils"
+)
+
+type TypedData struct {
+	types       map[string]TypeDefinition
+	primaryType string
+	domain      Domain
+	message     map[string]any
+	revision    *revision
+}
+
+// Types returns a copy of the TypedData's type definitions map
+func (td *TypedData) Types() map[string]TypeDefinition {
+	copyMap := make(map[string]TypeDefinition, len(td.types))
+	for k, v := range td.types {
+		copyMap[k] = v
+	}
+	return copyMap
+}
+
+// PrimaryType returns the primary type name of the TypedData
+func (td *TypedData) PrimaryType() string {
+	return td.primaryType
+}
+
+// Domain returns the domain information of the TypedData
+func (td *TypedData) Domain() Domain {
+	return td.domain
+}
+
+// Message returns a copy of the TypedData's message map
+func (td *TypedData) Message() map[string]any {
+	copyMap := make(map[string]any, len(td.message))
+	for k, v := range td.message {
+		copyMap[k] = v
+	}
+	return copyMap
+}
+
+// Revision returns the revision value of the TypedData
+func (td *TypedData) Revision() revision {
+	return *td.revision
+}
```

Comment:
I believe this is also performing a shallow copy of a reference type, see the above comment.

Reply:
Performing a deep copy of this should be straight forward given we know all the types ahead of time. But we should decide what to do in the above comment first